feat(products): Add barcode scanner to product form

ellpanda23 opened the pull request:
Integrates the html5-qrcode library to introduce a barcode scanning feature in the product creation and editing forms.

Initially, I attempted to integrate `html5-qrcode` as an NPM module. However, due to my limited experience with Filament's specific Vite integration for dynamically loaded modal content, I encountered challenges getting the compiled assets to load correctly. To ensure a stable and functional contribution, I opted for the more direct CDN-based approach, which works reliably. I am keen to receive feedback on how this could be better implemented with Vite in the future.

Key changes include:
- Added a new `generateBarcodeFormComponent` to create the barcode input field with a camera icon action.
- The action opens a modal containing the camera view for scanning.
- On successful scan, the barcode input is automatically populated, and the modal closes.
- The scanner is correctly stopped when the modal is closed, either automatically after a scan or manually by the user, preventing the camera from remaining active."

## What’s this PR about?

This pull request introduces a barcode scanning functionality to the product form, allowing users to populate the barcode field using their device's camera. This enhances user experience by speeding up data entry and reducing manual errors.

The implementation uses the html5-qrcode library, loaded via CDN, and is integrated as a modal action within a Filament form component.

---

## Key Changes:
Panel Provider (TenantPanelProvider.php): The html5-qrcode library is now loaded from the unpkg CDN via the panel's asset registration.

Product Form Trait (HasProductForm.php): A new generateBarcodeFormComponent method has been added. It creates the TextInput and a suffix action that triggers the scanner modal. The modal's cancel button is assigned a static ID for JavaScript targeting.

Scanner View (barcode-scanner.blade.php): A new Blade view was created to house the scanner's UI and the inline Alpine.js logic. This script handles:

Initializing the camera.

Populating the input field and dispatching an input event to notify Livewire upon a successful scan.

Programmatically clicking the close button to dismiss the modal.

Adding an event listener to the close button to ensure the scanner stops if the modal is closed manually.

## 🧾 Screenshots

<img width="620" height="119" alt="image" src="https://github.com/user-attachments/assets/73a35c16-b8ba-41cb-bcfa-0f7c73564017" />

<img width="537" height="541" alt="image" src="https://github.com/user-attachments/assets/2533f68b-1d73-4a26-9297-0a70199401f0" />
<img width="617" height="114" alt="image" src="https://github.com/user-attachments/assets/3c6e4faa-1c0d-4055-ab7c-d8ea18eb28ad" />

## How to Test:
1Navigate to the "Create Product" or "Edit Product" page.
2. Locate the "Barcode" input field.
3. Click the camera icon on the right side of the field. 
4. A modal with a camera view should appear. Grant camera permissions if prompted. 
5. Scan a standard barcode (e.g., EAN-13, CODE-128). 
6. Verify that upon successful scan, the modal closes automatically, and the barcode field is populated with the correct value. 
7. Re-open the modal and click the "Cancel" button. Verify that the modal closes and the camera deactivates.

---

__Note:__ This is my first-ever contribution to an open-source repository. I'm excited to learn and would greatly appreciate any feedback on my code, the commit structure, or the pull request itself. I'll be attentive to any comments or requested changes. Thank you for the opportunity!*

ellpanda23 committed:
feat(products): add barcode scanner to product form

Integrates the html5-qrcode library to introduce a barcode scanning feature in the product creation and editing forms.

Key changes include:
- Added a new `generateBarcodeFormComponent` to create the barcode input field with a camera icon action.
- The action opens a modal containing the camera view for scanning.
- On successful scan, the barcode input is automatically populated, and the modal closes.
- The scanner is correctly stopped when the modal is closed, either automatically after a scan or manually by the user, preventing the camera from remaining active."

ellpanda23 committed:
refactor(barcode-scanner): translate messages and comments to EnglishThe labels and messages in the barcode scanning component are updated to be in English. This includes changes to button texts, console messages, and comments in the code, improving language consistency in the application.

sheenazien8 commented:
Thank you so much for your contribution! 🎉
The implementation looks really good and I appreciate the effort, especially with handling the camera shutdown when the modal closes.


One thing I noticed: currently the scanner is only available in the product form. It would be even more useful if we also support scanning directly in the POS transaction page (so users can scan products when making a sale).
@ellpanda23

sheenazien8 commented:
> Thank you so much for your contribution! 🎉 The implementation looks really good and I appreciate the effort, especially with handling the camera shutdown when the modal closes.
> 
> One thing I noticed: currently the scanner is only available in the product form. It would be even more useful if we also support scanning directly in the POS transaction page (so users can scan products when making a sale). @ellpanda23

you can check the existing code for that is in https://github.com/lakasir/lakasir/blob/b39171258cb1a2e11214cfe23c58ba577b06fb54/resources/views/filament/tenant/pages/cashier.blade.php#L566

ellpanda23 commented:
Hi @sheenazien8,

Thank you so much for the positive feedback! I'm really glad you like the implementation.

That's an excellent suggestion to add the scanner to the POS page as well. I completely agree it would be a very useful feature for processing sales quickly. Thank you for pointing me to the `cashier.blade.php` file and the existing `keypress` event listener – that's a perfect starting point.

I'll start working on integrating it right away. I will push the new commits to this branch and update the PR once it's ready for another review.

Thanks again for your guidance!

ellpanda23 committed:
refactor(cashier): improve code structure and readability

Changes were made to the cashier view to optimize code readability and structure. Styles were adjusted and the organization of elements was improved, including the implementation of a QR code scanner in a modal. Additionally, some minor bugs were fixed and event handling for barcode scanning was improved.

ellpanda23 committed:
refactor(cashier): translate comments and messages to English

Comments and messages in the cashier view component are updated to be in English, improving language consistency across the application. Adjustments were made to the code structure and the state management of the QR scanner was optimized.

ellpanda23 commented:
## feat: Implement Advanced Camera Barcode Scanner for Cashier Interface

### Summary
This pull request introduces a robust and user-friendly camera-based barcode scanning feature to the cashier/POS interface. It allows users to scan product barcodes directly with their device's camera, providing a seamless alternative to a physical scanner.

The implementation is designed for a fast-paced retail environment, featuring continuous scanning, immediate UI feedback, performance optimizations, and several bug fixes to ensure stability with Livewire's lifecycle.

### Key Features & Changes ✨
📷 Camera Scanner Integration:

Integrates the html5-qrcode library to enable scanning via the device's camera.

A new QR code icon has been added to the cashier header to launch the scanner modal.

The library automatically provides a camera selection dropdown if multiple cameras (e.g., front/rear) are detected.

🔄 Continuous Scanning Mode:

The scanner modal remains open after a successful scan, allowing the cashier to scan multiple items in rapid succession without reopening the modal.

⏳ Scan Cooldown (Throttling):

A 1-second cooldown period is enforced after each successful scan. This prevents a single barcode from being scanned and added multiple times if held in front of the camera for too long.

⚙️ Asynchronous UI Feedback (Loading Spinner):

After a barcode is scanned, the camera view is temporarily replaced by a loading spinner.

This provides clear visual feedback that the system is processing the product.

The camera view reappears automatically once the Livewire backend call is complete, ready for the next scan.

🚀 Performance & Stability Fixes:

DOM Preservation: Implemented wire:ignore on the scanner's container to prevent Livewire from removing the camera view during component re-renders. This was a critical fix to enable continuous scanning.

Optimized Barcode Detection: The scanner is now configured with formatsToSupport to look only for specific barcode types (EAN, UPC, Code128, etc.), improving detection speed and accuracy.

Robust State Management: Refactored the JavaScript to use a more reliable method (element.closest()) for interacting with the Alpine.js component, fixing a null reference error.

Clean Shutdown: The scanner is now reliably stopped (releasing the camera) when the user manually closes the modal.

### Visual Demonstration
Here’s a quick demonstration of the new workflow: scanning multiple items, the loading spinner appearing for each, and the cart updating in real-time.

### How to Test This PR 🧪
1. Navigate to the Cashier/POS page.

2. Click the new QR Code icon in the header section.. 
<img width="835" height="216" alt="image" src="https://github.com/user-attachments/assets/f4688d0f-27b5-486b-9af5-b974083c2924" />

3. The scanner modal should open, and your device camera should activate.
<img width="812" height="772" alt="image" src="https://github.com/user-attachments/assets/8f8a21d5-6254-484c-bfbe-182ecc2c3067" />

4. Scan a valid product barcode.

5. ✔️ Verify: The camera view is replaced by a loading spinner.
<img width="578" height="511" alt="image" src="https://github.com/user-attachments/assets/ef3ea577-e343-41b7-829e-eada88ee508a" />

6. ✔️ Verify: The product is added to the cart on the main page.
<img width="855" height="290" alt="image" src="https://github.com/user-attachments/assets/cf76641c-36cd-499f-b3d0-1da7a599fe37" />

7. ✔️ Verify: The loading spinner disappears, and the camera view returns, ready for the next scan.

8. Immediately try to scan the same barcode again within one second.

9. ✔️ Verify: The product is not added again due to the cooldown mechanism. After one second passes, it can be scanned again.

10. Scan a different product barcode.

11. ✔️ Verify: The second product is successfully added to the cart.

12. Click the "Close" button on the modal.

13. ✔️ Verify: The modal closes, and your device's camera light turns off.

@sheenazien8

sheenazien8 commented:
Hi @ellpanda23 ,

Thank you so much for this contribution — I really appreciate the effort you put in to bring barcode scanning into the product form! 🙏


I’ll pull this branch and test it out locally first to see how it works in our environment.

thanks! 🚀

ellpanda23 commented:
Important: To test the camera barcode scanner feature, you must serve the application over a secure connection (HTTPS).
Modern web browsers enforce a "Secure Contexts" policy for privacy-sensitive APIs, which includes getUserMedia (the API for accessing the camera). This security measure prevents unencrypted sites (HTTP) from accessing a user's camera, protecting them from potential man-in-the-middle attacks.
How to Test Locally
• If you are using a local development environment like Laravel Valet or Herd, they automatically serve sites with a .test domain over HTTPS. Simply navigate to https://your-site.test.
• If you are using Vite, you may need to enable its HTTPS flag.
• Accessing the site via a non-secure http:// URL (unless it's http://localhost) will result in the camera failing to initialize, and an error will be logged in the browser console.

sheenazien8 commented:
> Important: To test the camera barcode scanner feature, you must serve the application over a secure connection (HTTPS). Modern web browsers enforce a "Secure Contexts" policy for privacy-sensitive APIs, which includes getUserMedia (the API for accessing the camera). This security measure prevents unencrypted sites (HTTP) from accessing a user's camera, protecting them from potential man-in-the-middle attacks. How to Test Locally • If you are using a local development environment like Laravel Valet or Herd, they automatically serve sites with a .test domain over HTTPS. Simply navigate to https://your-site.test. • If you are using Vite, you may need to enable its HTTPS flag. • Accessing the site via a non-secure http:// URL (unless it's http://localhost) will result in the camera failing to initialize, and an error will be logged in the browser console.

sweet, thanks for your advice
